presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -1046,28 +1055,167 @@ public QueryInfo updateQueryInfo(Optional<StageInfo> stageInfo)
      * Remove large objects from the query info object graph, e.g : plan, stats, stage summaries, failed attempts
      * Used when pruning expired queries from the state machine
      */
-    public void pruneQueryInfo()
+    public void pruneQueryInfo(PruneLevel level)
     {
         Optional<QueryInfo> finalInfo = finalQueryInfo.get();
         if (!finalInfo.isPresent() || !finalInfo.get().getOutputStage().isPresent()) {
             return;
         }
 
         QueryInfo queryInfo = finalInfo.get();
+        QueryInfo prunedQueryInfo;
+        switch (level) {
+            case FINISHED:
+                // no longer needed in the session after query finishes
+                session.getPlanNodeStatsMap().clear();
+                session.getPlanNodeCostMap().clear();
+                // inputs contain some statistics which should be cleared
+                inputs.getAndUpdate(QueryStateMachine::pruneInputHistograms);
+                // query listeners maintain state in their arguments which holds
+                // onto plan nodes and statistics. Since finalQueryInfo was
+                // already set it should be in a terminal state and be safe to
+                // clear the listeners.
+                finalQueryInfo.clearEventListeners();
+                planStatsAndCosts.getAndUpdate(stats -> Optional.ofNullable(stats)
+                        .map(QueryStateMachine::pruneHistogramsFromStatsAndCosts)
+                        .orElse(null));
+                prunedQueryInfo = pruneFinishedQueryInfo(queryInfo, inputs.get());
+                break;
+            case EXPIRED:
+                prunedQueryInfo = pruneExpiredQueryInfo(queryInfo, getMemoryPool());
+                break;
+            default:
+                prunedQueryInfo = queryInfo;
```

Comment:
nit: Is it better to throw an exception here? As it shouldn't happen at all.

Reply:
Since this isn't the only place that `PruneLevel` is used, the thinking was that if someone were to add another enum entry in the future, it may not necessarily be desirable for this method to throw an exception because the implementation isn't entirely done in this method. 

For example, `PruneLevel` is also used here: https://github.com/prestodb/presto/blob/735877a34c1541af1a5900536633bdf60d124ace/presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java#L750-L762

Reply:
IMO, throw an exception by default can make the responsibilities more clearer, even considering the scenario you mentioned above.

For example, if a coder add another enum entry in `PruneLevel` in future, but explicitly do not want to support it in `QueryStateMachine.pruneQueryInfo(level)` for some reasonable reasons, it's this coder's responsibility to make sure not invoking the function `stateMachine.pruneQueryInfo(level)` with the newly added unsupported level, or he/she will get a explicitly exception.

Meanwhile, if a coder add another enum entry in `PruneLevel` in future, but forget to support it in `QueryStateMachine.pruneQueryInfo(level)`, he/she will receive a reminder immediately when invoking `stateMachine.pruneQueryInfo(level)` with the newly added level. It's better than being ignored silently, which may result in not being able to detect this issue for a long while.

What's your opinion?

Reply:
I can understand your reasoning. It has been updated to throw an exception instead